### ISSUE TITLE
Accessing Database and MYCLUSTER_SECRET from Task

### ISSUE THREAD

**Issue by gidsg** [aws/copilot-cli#4989]
I have a Copilot managed database which is exposed to a Backend Web Service with its credentials automatically exposed in an env var i.e.: `MYCLUSTER_SECRET` as described in: https://aws.github.io/copilot-cli/docs/developing/storage/#database-and-artifacts. I would like to access the database in a task with `copilot task run` however it appears the `MYCLUSTER_SECRET` env var is not available to the task e.g.
```
$ copilot task run --app app-name --env env-name --follow --command "flask db upgrade" 
...
KeyError: 'MYCLUSTER_SECRET'
```
How can I expose the database secret to the task and also provide it with read/write access to the database instance?

My copilot config is also available at https://github.com/communitiesuk/funding-service-design-post-award-data-store/tree/main/copilot for reference.

**Comment by huanjani** [comment]
Hello, @gidsg! 
Have you tried leveraging the `--secrets` flag? (https://aws.github.io/copilot-cli/docs/commands/task-run/) 
Here's an outdated issue that can give you some context: https://github.com/aws/copilot-cli/issues/2641

**Comment by gidsg** [comment]
@huanjani 
> Have you tried leveraging the --secrets flag? (https://aws.github.io/copilot-cli/docs/commands/task-run/)

I tried this now, but unfortunately am hitting errors and have been unable to get this to work.
The following is unclear after looking at the documentation:
* The `MYCLUSTER_SECRET` copilot generated storage credential is stored in Secrets Manager whereas user defined secrets are stored in Parameter Store. Do you know if the `--secrets` searches both Secrets Manager and Parameter Store?
* Do you know if the `--secrets` flag is expecting the _Secret name_ from secrets manager or some other value?
* When I use the `--secrets` flag I am getting the following permissions error:
```
✘ It looks like your task is not able to pull the secrets.
Did you tag your secrets with the "copilot-application" and "copilot-environment" tags?
✘ run task task-name-app-name-svc-name: run task task-name-app-name-svc-name: task 3f1fb099: ResourceInitializationError: unable to pull secrets or registry auth: execution resource retrieval failed: unable to retrieve secrets from ssm: service call has been retried 1 time(s): AccessDeniedException: User: arn:aws:sts::REDACTED:assumed-role/task-task-name-p-DefaultExecutionRole-REDACTED is not authorized to perform: ssm:GetParameters on resource: arn:aws:ssm:eu-west-2:REDACTED:parameter/MYCLUSTER_SECRET because no identity-based policy allows the ssm:GetParameters action status code: 400, request id: a680dc92-42fd-49e0-98fe-b026f985bf74
```
* It looks like from the above error it is searching Parameter Store rather than Secrets Manager?

**Comment by KollaAdithya** [comment]
Hey @gidsg !
1. `--secrets` flag searches for both Parameter store and Secrets Manager.
2. If you creating secret using `copilot secret init` then Copilot by default will store the secret in secrets manager and tags it with `copilot-application` and `copilot-environment`.
3. By looking at the error, I would assume you are using `--secrets` flag with user defined secret. Usually this error happens when the secret is not tagged with `copilot-application` and `copilot-environment`.
Can you tag your secret with tag as `copilot-application` and value = `Replace with your app name`
tag `copilot-environment` and value = `Replace with your env name`

You can refer to the secrets guide https://aws.github.io/copilot-cli/docs/developing/secrets/

**Comment by gidsg** [comment]
@KollaAdithya @huanjani 

> By looking at the error, I would assume you are using --secrets flag with user defined secret.

This secret was not created by `copilot secret init` it was autocreated by `copilot storage init` as described in https://aws.github.io/copilot-cli/docs/developing/storage/#database-and-artifacts
...
> This will create an RDS Aurora Serverless v2 cluster that uses PostgreSQL engine with a database named my_db. An environment variable named MYCLUSTER_SECRET is injected into your workload as a JSON string. The fields are 'host', 'port', 'dbname', 'username', 'password', 'dbClusterIdentifier' and 'engine'.

 > Usually this error happens when the secret is not tagged with copilot-application and copilot-environment.

The secret is tagged as such e.g. 
![Screenshot 2023-06-20 at 17 04 31](https://github.com/aws/copilot-cli/assets/1764158/4427c181-b0d5-41a0-9c8d-ad62c60ab8c1)

Please advise how I may access the secrets, or are there extra constraints as it was created by `copilot storage init`

**Comment by KollaAdithya** [comment]
Hi @gidsg ! I tried to reproduce your use case. I am able to use `--secrets` flag with `copilot task run` command with out any error.

1. This is the secret which Copilot automatically creates for you when you run `copilot storage init` in `AWS secrets manager`.

![Screenshot 2023-06-20 at 7 53 03 PM](https://github.com/aws/copilot-cli/assets/71282729/3a43348e-4b25-4bef-8e3f-849d081b4b38)


2.  Then I run the  below command with `--secrets` with the `ARN` of the secret created by `copilot storage init`. I am able to successfully deploy the task with out any error.
```
copilot task run --secrets backendclusterAuroraSecret=arn:aws:secretsmanager:us-east-1:535307839156:secret:backendclusterAuroraSecret-x3QmTTnxXtoS-go5GZc --image 535307839156.dkr.ecr.us-east-1.amazonaws.com/hyatt/backend:latest
```
![Screenshot 2023-06-20 at 7 56 39 PM](https://github.com/aws/copilot-cli/assets/71282729/2fa8d349-5524-4f37-b554-40901415bcc5)

**Comment by gidsg** [comment]
@KollaAdithya thanks can confirm using the full ARN worked for me to access the secret. In addition I had to specify a security group to get access to the database.

Some feedback on this process:

* It was not clear from the documentation that the full ARN was required rather than the secret name
* To get read-write access to the database I also had to add a security group
* It would be good to be able to get these ARN and Security Group values from the copilot CLI rather than having to fetch them manually from the console and having to pass these in per-environment. Is this possible already and I've missed this option?
* It would be a good improvement if a task could inherit a service context. i.e. the environment and permissions you get when running `copilot svc exec` but as a one-off task. This would be similar to the `run-task` option provided by Cloudfoundry CLI e.g. https://docs.cloudfoundry.org/devguide/using-tasks.html

**Comment by KollaAdithya** [comment]
Hello @gidsg !

So if the pain point is only to fetch the `secret ARN` and `security group` from the Console manually and then pass it to `task run`.

Can you do `copilot task run --generate-cmd --generate-cmd <application>/<environment>/<service name>`. This will generate a command prefilled with all the flag values.

**Comment by gidsg** [comment]
@KollaAdithya 

> Can you do copilot task run --generate-cmd --generate-cmd <application>/<environment>/<service name>. This will generate a command prefilled with all the flag values.

Thanks for the suggestion, yes, this approach worked for us. Our script is publicly accessible at: https://github.com/communitiesuk/funding-service-design-post-award-data-store/blob/03ad493af40772d7323c38f3211cb0219ca6e62a/scripts/migration-task-script.sh in case useful for others.

**Comment by huanjani** [comment]
The docs update has been released in v1.29.0: https://github.com/aws/copilot-cli/releases/tag/v1.29.0!
